### DISCUSSION TITLE
How to add this equation to the kernel?

### DISCUSSION THREAD

**Discussion by YuhangNIU** [idaholab/moose#20353] [Q&A General] [answered]
Hello all,

I have a small question. Suppose that one-dimensional geometry is divided into _**n**_ elements. The equation in element _i_ (0<_i_<n) is:
![image](https://user-images.githubusercontent.com/80933114/154967848-9dfbb9d1-7bb1-4dbb-aff4-42107e753d7d.png)
Could anyone kindly tell me how to add this problem to the kernel?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Can you write this [a,b,c] stencil as a gradient / laplacian / other known operator or is it arbitrary?
Do you want to use a finite volume or finite element discretization?

For the d term, you can use:
https://mooseframework.inl.gov/source/kernels/BodyForce.html

Guillaume

**Reply by YuhangNIU** [reply]
Hi,

Thanks for your reply. These parameters (a, b, c, d) are constant. By the way, the variable in this equation is W. 
I think any discretization is acceptable, as long as it can be expressed in this equation.  I don't know how the kernel can be constructed.

Thank you,
Yuhang

**Reply by hugary1995** [reply]
This is neither FE nor FV, but looks more like finite differencing. What's your motivation of using moose?

**Reply by YuhangNIU** [reply]
I want to develop a similar equation based on moose.
Does it mean that moose can not solve this equation?

**Reply by hugary1995** [reply]
Yuhang,

If you are okay with any discretization, then please show us the equation you want to solve for -- the equation before discretization, not the equation shown in your original post. Then we can tell you which kernels to use.

**Reply by YuhangNIU** [reply]
![image](https://user-images.githubusercontent.com/80933114/155248576-2d8e2ebd-94a3-4bbf-9d74-b65c6663920f.png)
![image](https://user-images.githubusercontent.com/80933114/155248596-3e42ac2b-215c-40c4-8ec0-ffff7204df67.png)
![image](https://user-images.githubusercontent.com/80933114/155248611-6e97457c-31f5-41d4-9f77-15a3f603d543.png)

I just want to solve this equation, but I do not know how to express this equation before discretization. Can it be solved with MOOSE?

**Reply by GiudGiud** [reply]
It can be solved by MOOSE a priori, but it would be easier with the undiscretized form of the equation.

A few questions,
l and i are indexes for two spatial dimensions?
What are the variables you want to solve for here? Since you have two PDEs, we cant have too many variables, otherwise the problem will be underdetermined. I count rho, h, W, but seems WBE, Q, T, P and C could also be variables? Or are they constants?

**Reply by YuhangNIU** [reply]
Thanks for your reply. 
The index of l means that this equation is for liquid, so it is nothing special. And i  represents the index of 1D elements.
The variables of these two PDEs are W and h, so other parameters can be regarded as constant.

**Reply by YuhangNIU** [reply]
Besides, if W < 0, the second term on the right of Eq.2 can be expressed as: W(i)*h(i+1). However, the h value of i + 1 element (_u[qp+1]) is unknown in this time step, resulting in divergence. How to solve this problem？

**Reply by GiudGiud** [reply]
> Thanks for your reply. The index of l means that this equation is for liquid, so it is nothing special. And i represents the index of 1D elements. The variables of these two PDEs are W and h, so other parameters can be regarded as constant.

so the d(rho V/m) / dt term is 0? if all of these are constant

If the XX_i-1 - XX_i terms can be written as gradients, then this is solvable. If not, it's a little problematic that it's only on one side of the 1D element.

**Reply by YuhangNIU** [reply]
Yes, the d(rho V/m) / dt term is 0. 
I was confused that the gradient is defined as d(u) / dx in 1D element, but the Eq.(1) and Eq.(2) only can be written as d(u). 
I think it might be written as grad(u) * dx, but how to define dx in moose?

**Reply by GiudGiud** [reply]
dx will be the size of the mesh cells. In a 1D mesh it s just the length of an element.
Your typical gradient wont be one-sided like this though, which is why we are insisting about the undiscretized version of the equation. 

I ll ask @lindsayad to confirm that this isnt easier with DG kernels, but otherwise, I think it'd be easier to set this problem up in Python.

**Reply by YuhangNIU** [reply]
Thanks for your kindness.

**Reply by lindsayad** [reply]
It would be a lot easier to comment if I saw the strong form of the equation, e.g. including spatial derivative operators like gradient and divergence. It takes too much time for me to grok the form as it's currently given
